src/memory_agent_action.h

```diff
@@ -10,31 +10,33 @@
 #include "log.h"
 #include "utils.h"
 #include "cancellation_manager.h"
+#include "progress_manager.h"
 
 #define MEMORY_AGENT_INTERRUPTED_ERROR static_cast<jvmtiError>(999)
 
-struct CallbackWrapperData {
-    CallbackWrapperData(void *callback, void *userData, const CancellationManager *manager) :
-            callback(callback), userData(userData), manager(manager) {
+template<typename RESULT_TYPE, typename... ARGS_TYPES>
+class MemoryAgentAction : public CancellationManager, public ProgressManager {
```

Comment:
Why did you decide to prefer inheritance over aggregation here? I'd say aggregation more natural here